### PR TITLE
Enabled a way to buffer request body in formatters

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs [hunk 1]
```diff
@@ -75,6 +75,11 @@ public MvcOptions()
         public FormatterCollection<IInputFormatter> InputFormatters { get; }
 
         /// <summary>
+        /// Gets or sets the flag to buffer the request body in formatters. Default is <c>true</c>.
+        /// </summary>
+        public bool BufferRequestBodyInFormatters { get; set; } = true;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs]: This should be named in such a way that it's `false` by default. It's very surprising when a flag isn't false by default.

Maybe `SupportInputFormatterBuffering`?

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs [hunk 2]
```diff
@@ -98,6 +127,18 @@ public class JsonInputFormatter : TextInputFormatter
             }
 
             var request = context.HttpContext.Request;
+
+            if (_bufferRequestBody)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs]: We should get `request.Body.CanSeek` here. Then if the body is already buffered we don't need to do anything.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs [hunk 3]
```diff
@@ -98,6 +127,18 @@ public class JsonInputFormatter : TextInputFormatter
...
+            {
+                // JSON.Net does synchronous reads. In order to avoid blocking on the stream, we asynchronously 
+                // read everything into a buffer, and then seek back to the beginning. 
+                BufferingHelper.EnableRewind(request);
+                Debug.Assert(request.Body.CanSeek);
+
+                await request.Body.CopyToAsync(Stream.Null);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs]: Use https://github.com/aspnet/HttpAbstractions/blob/87cd79d6fc54bb4abf07c1e380cd7a9498a78612/src/Microsoft.AspNetCore.WebUtilities/StreamHelperExtensions.cs#L15

File: test/Microsoft.AspNetCore.Mvc.Core.Test/MvcOptionsTest.cs [hunk 4]
```diff
@@ -18,5 +18,15 @@ public void MaxValidationError_ThrowsIfValueIsOutOfRange()
             var ex = Assert.Throws<ArgumentOutOfRangeException>(() => options.MaxModelValidationErrors = -1);
             Assert.Equal("value", ex.ParamName);
         }
+
+        [Fact]
+        public void BufferingRequestBody_EnabledByDefault_ForFormatters()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/MvcOptionsTest.cs]: meh

File: src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs]: You meant `SuppresstInputFormatterBuffering = false`?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs]: Yes.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs [hunk 5]
```diff
@@ -98,6 +128,22 @@ public class JsonInputFormatter : TextInputFormatter
             }
 
             var request = context.HttpContext.Request;
+
+            if (request.Body.CanSeek)
+            {
+                request.Body.Seek(0L, SeekOrigin.Begin);
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs]: 🆙 📅  Should we do this or the layer which buffered the request should rewind it back?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs]: We should do this. 

Notice the follow is actually:
1. Enable buffering
2. Read to the end
3. Rewind to the beginning

We do 2, so we should do 3.

Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs]: Sure, sounds good, so I will leave it as it is.

**Comment by kichalla** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs]: This call to seek at L134 isn't needed. We do need the one at L144.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlDataContractSerializerInputFormatter.cs [hunk 6]
```diff
@@ -99,6 +114,22 @@ public override Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterCo
             }
 
             var request = context.HttpContext.Request;
+
+            if (request.Body.CanSeek)
+            {
+                request.Body.Seek(0L, SeekOrigin.Begin);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlDataContractSerializerInputFormatter.cs]: This isn't needed.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlSerializerInputFormatter.cs [hunk 7]
```diff
@@ -80,6 +95,22 @@ public XmlDictionaryReaderQuotas XmlDictionaryReaderQuotas
             }
 
             var request = context.HttpContext.Request;
+
+            if (request.Body.CanSeek)
+            {
+                request.Body.Seek(0L, SeekOrigin.Begin);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Xml/XmlSerializerInputFormatter.cs]: This isn't needed.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs [hunk 8]
```diff
@@ -98,6 +128,22 @@ public class JsonInputFormatter : TextInputFormatter
...
+            }
+            else if (!_suppressInputFormatterBuffering)
+            {
+                // JSON.Net does synchronous reads. In order to avoid blocking on the stream, we asynchronously 
+                // read everything into a buffer, and then seek back to the beginning. 
+                BufferingHelper.EnableRewind(request);
+                Debug.Assert(request.Body.CanSeek);
+
+                await request.Body.DrainAsync(context.HttpContext.RequestAborted);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonInputFormatter.cs]: I don't think we should pass a cancellation token here. MVC doesn't have any other behavior where we cancel based on request aborted.

**Comment by kichalla** [comment]
🆙 📅

**Comment by kichalla** [comment]
f1dd475eae83ca32950e82d5ac2facfe1eaad42e

**Pull request closed without merging**